Add protos for RBAC support

liminw opened the pull request:
Goals:
* Provide RBAC authorization in Envoy.
* Provide simple user interface for basic use cases, while allow flexibility. 
* Compatible with Istio RBAC (https://istio.io/docs/concepts/security/rbac.html).

Proto files: 
* rbac.proto is mainly moved from https://github.com/istio/api/blob/master/rbac/v1alpha1/rbac.proto. It focuses on the RBAC interface provided to user. 
* local_rbac.proto specifies the filter configuration received by Envoy.

The proto here defines the full RBAC interface. Implementation will starts with basic support (without "constraints" and "custom properties") and grow into fully support RBAC. 

Background reads: "Istio RBAC on Envoy" design doc (https://docs.google.com/document/d/1XkN9ls-ZDISz2uXGN20RaIre63DkJJOJNUOzcKCxbWA).

liminw commented:
/cc. @mattklein123 @rodaine @danielhochman

liminw commented:
/cc @lizan @yangminzhu @JimmyCYJ

rodaine commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,180 @@
+  string group = 2;
+
+  // Optional. The set of properties that identify the subject.
+  map<string, string> properties = 3;
```
could we reuse the `Constraint` message here instead?

rodaine replied:
Or conversely for the constraints field above, could also use `map<string, string>` instead?

rodaine replied:
Or, why not use the AttributeContext directly? It seems to explicitly define the identifiers we want here.

mattklein123 replied:
+1, let's make sure to use concrete attributes vs. abstract strings, and let's reuse the ones from extauth if possible.

rodaine commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,180 @@
+message RoleRef {
+  // Required. The type of the role being referenced.
+  // Currently, "ServiceRole" is the only supported value for "kind".
+  string kind = 1 [(validate.rules).string.const = "ServiceRole"];
```
can we use an enum here?

liminw replied:
Based on API style naming convention (https://cloud.google.com/apis/design/naming_convention#enum_names), "Enum values must use CAPITALIZED_NAMES_WITH_UNDERSCORES". But I want to use "ServiceRole" constant here, because it needs to match ServiceRole definition "kind: ServiceRole".

rodaine replied:
The name is arbitrary (when serialized enums are represented as integers). We aren't doing string comparison between these are we?

lizan replied:
You don't even need this, you can always add this later or upgrade to `oneof` without breaking backward compatibility. I'd quote [STYLE.md](https://github.com/envoyproxy/envoy/blob/master/STYLE.md) again that prefers `oneof` than enum + string.

liminw replied:
ok, let me remove "kind" for now. I can add it later if needed.

rodaine commented on envoy/api/v2/auth/local_rbac.proto:
```diff
@@ -0,0 +1,67 @@
+
+import "envoy/api/v2/auth/rbac.proto";
+
+// LocalRBAC specifies Istio RBAC policies that apply to the local RBAC engine
```
Do we want to keep the Istio references throughout, or just link back to the existing Istio documentation for reference?

rodaine commented on envoy/api/v2/auth/local_rbac.proto:
```diff
@@ -0,0 +1,67 @@
+//       bindings:
+//       - subjects:
+//         - user: “cluster.local/ns/default/sa/admin”
+//         roleRef:
```
since each `Policy` holds its own `ServiceRoleBinding`s, what is the need for the `RoleRef`? Are there cases where the binding is independent of the `Policy`?

liminw replied:
I made roleRef an optional field in ServiceRoleBinding. This way, when used in filter config, roleRef part can be skipped. RoleRef is needed when a user defines a ServiceRoleBinding object. I updated the example here and removed "roleRef" section.

rodaine replied:
So you're saying when the Bindings is independent of the Policy? When does that happen?

liminw replied:
When a user defines ServiceRoleBinding object, which is independent of ServiceRole. See the example of ServiceRoleBinding in rbac.proto. RoleRef is included here for completeness of the proto. 

Does the user interaction happen in control plane only? Is there a case an Envoy is deployed for a service without a control plane? In that case, a user may want to enter ServiceRole and ServiceRoleBinding independently, and the underlying tooling associate a binding to a role based on RoleRef. 

I think @mattklein123 asked the same question.

mattklein123 commented:
Quick note (not full review). As part of this change, can we kill https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/auth/auth.proto? It's currently not used and this functionality should be a full replacement for it.

lizan commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,180 @@
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // Exact match, prefix match, and suffix match are supported for paths.
+  // For example, the path "/books/review" matches
```
Extract the combination of (exact match, prefix match, suffix match) to a message with `oneof` like [this](https://github.com/envoyproxy/data-plane-api/blob/master/STYLE.md) (scroll down for oneof)? It is more self descriptive than a comment.

lizan replied:
Yes I understand this will change the language slightly though I think it's worth. The idea is keep this extensible and self describing. For example, someone may add regex later but you don't want always do regex.

rodaine commented on envoy/api/v2/auth/BUILD:
```diff
@@ -50,3 +50,31 @@ api_go_proto_library(
     ],
 )
+
+api_proto_library(
```
So the RBAC filters will exist both at the listener (to support non-http binary protocols) as well as the router level, so placing them solely within the scope of the HTTP namespace would be misleading. Is there a better place you might suggest?

lizan replied:
Then create `envoy/config/rbac/v2alpha`?

Per [style guide](https://github.com/envoyproxy/data-plane-api/blob/master/STYLE.md#package-organization) here doesn't sound right.

mattklein123 replied:
I agree. I originally talked with @liminw and thought this should go in config. I would go with the path that @lizan suggests.

htuch commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,181 @@
+// object.
+message ServiceRoleBinding {
+  // Required. List of subjects that are assigned the ServiceRole object.
+  repeated Subject subjects = 1;
```
Can this be empty? Can we add a constrain that the list length is non-zero?

liminw replied:
Added a min_items constraint.

mattklein123 commented:
> After a quick scan, my main feedback is that most of the docs can't be read independently of Istio's APIs.

@liminw I agree that we need to purge all mention of Istio from this PR. I still haven't reviewed this and will do so by EOD but I wanted to +1 @htuch comment here. This inline RBAC API should be completely independent of Istio.

mattklein123 commented on envoy/api/v2/auth/local_rbac.proto:
```diff
@@ -0,0 +1,64 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
```
I think we can remove draft stuff as long as we go with v2alpha (same elsewhere). Doesn't really matter though as long as we do v2alpha.

mattklein123 commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,181 @@
+  // If set to ["*"] or not specified, it applies to any method.
+  repeated string methods = 3;
+
+  // Definition of a custom constraint. The key in the constraint refers to an Istio attribute name
```
For now my preference is to remove any abstract constraints. I don't see how they make sense for local execution in the proxy?

mattklein123 commented on envoy/api/v2/auth/rbac.proto:
```diff
@@ -0,0 +1,181 @@
+}
+
+// RoleRef refers to a role object.
+message RoleRef {
```
I think I'm probably missing something, but how will this information actually be used by the inline filter? Logging? Settings role headers? If so can we document more clearly?

liminw commented on envoy/api/v2/auth/local_rbac.proto:
```diff
@@ -0,0 +1,67 @@
+//    },
+//  ]
+message LocalRBAC {
+  // A list of ServicePolicy. One per service.
```
Changed it to a map.

liminw commented:
@mattklein123 @rodaine @htuch @lizan @yangminzhu Thanks for the reviews!

I made the following major changes:
* Removed all references to Istio.
* Moved the files to envoy/config/rbac/v2alpha.
* Removed all comments about abstract constraints. 

Question:
I referenced a couple of time to //envoy/service/auth/v2/attribute_context.proto. We will use AttributeContext defined in the proto in RBAC implementation. Do you think it makes sense to move this proto to rbac folder, or some common places that both rbac and external_auth can use?

liminw commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,171 @@
+//  For example, destination.labels["app"], request.http.headers["version"].
+//
+// Below is an example of ServiceRole object "product-viewer", which has "read" ("GET" and "HEAD")
+// access to "products.default.svc.cluster.local" service at versions "v1" and "v2". "path" is not
```
Fixed.

liminw commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,171 @@
+//    user: source.principal
+//    group is currently undefined. It is reserved to be used in the future.
+//
+// Custom property name the attributes defined in AttributeContext
```
This part of text is removed.

liminw commented on envoy/config/rbac/v2alpha/local_rbac.proto:
```diff
@@ -0,0 +1,63 @@
+syntax = "proto3";
+
+package envoy.api.v2.auth;
```
Fixed.

liminw commented:
@mattklein123 @rodaine I was trying to remove https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/auth/auth.proto. It seems that it is currently used in https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/route/route.proto. Can you check to make sure that you want to remove all the usage of auth.proto? I do not have knowledge how this proto is used today, and want to double confirm with you.

mattklein123 commented:
@liminw the proto is not actually used in Envoy code. Go ahead and just remove from route.proto and just reserve the field that it was in (per the style.md docs). Thank you!

liminw commented:
@mattklein123 To answer your question on if RBAC covers all the functionality provided by existing auth.proto: RBAC defines rules based on attributes. The existing proto does access control on CertificateValidationContext. As long as we have all the relevant attributes, RBAC should be able to cover all the use cases. 

For the basic use cases like whitelisting calling service accounts for a service, you define a role with service name and allowed verbs, and list all the service accounts in the Binding. And it supports various AND/OR logics in Role and Binding definitions. In control plane, users can define roles that apply to all or a set of services (like the "service-admin" role and "tester" role examples https://istio.io/docs/concepts/security/rbac.html#servicerole).  

One thing keeping in mind is that, RBAC is deny by default (like many other authorization systems, e.g., K8S RBAC, Google Cloud IAM, AWS IAM). Users need to explicitly define RBAC policies to allow access. For the use case that you only want to restrict access for a few services, you can enable RBAC only for those services. The control to enable RBAC globally or only to a few services is normally configured in control plane (example for Istio: https://docs.google.com/document/d/1XkN9ls-ZDISz2uXGN20RaIre63DkJJOJNUOzcKCxbWA/edit#heading=h.lkkwo6ehmrav).

liminw commented:
@mattklein123 @rodaine @htuch @lizan @yangminzhu I think I have addressed all your comments. Please take a look when you get a chance. Thanks!

htuch commented on envoy/config/rbac/v2alpha/local_rbac.proto:
```diff
@@ -0,0 +1,63 @@
+//
+// Here is an example of local RBAC policy, which contains policies for a single service.
+// {
+//   "products.default.svc.cluster.local": {
```
Can you use pure YAML in your examples without superfluous JS braces? I think this will make it a bit easier to read.

liminw replied:
Will do. Thanks.

liminw replied:
Done. I moved this example to README.md.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+  // Optional. A list of HTTP verbs (e.g., "GET", "POST").
+  // It is ignored in gRPC case because the value is always "POST".
+  // If set to ["*"] or not specified, it applies to any verb.
+  repeated string verbs = 3;
```
This should probably be an enum.

liminw replied:
The most commonly used HTTP verbs are POST, GET, PUT, PATCH, and DELETE. But there are some uncommonly used ones, like HEAD, OPTIONS, TRACE, CONNECT. Customers may want to use their custom HTTP verbs, like PUBLISH. As long as the verb in policy matches the verb in the request, we will allow the request. For this reason, I do not want to limit the HTTP verbs users can use.

wora replied:
AttributeContext uses the term "method", not "verb".

liminw replied:
I previously used the term "method", but I changed it to "verb" because I was trying to align with the terminology used in Kubernetes RBAC (https://kubernetes.io/docs/admin/authorization/rbac/).

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+//           values: ["v1", "v2"]
+//
+// A ServiceRoleBinding specification includes two parts:
+// * "roleRef" refers to a ServiceRole object.
```
s/roleRef/role_ref/

liminw replied:
Done.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+//   path: request.http.path
+//
+//  In addition, users can specify "constraints" on any attributes defined in AttributeContext
+//  (//envoy/service/auth/v2/attribute_context.proto).
```
These links to proto files would benefit from being `:api:` references (here and elsewhere).

liminw replied:
Done.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+// * "roleRef" refers to a ServiceRole object.
+// * A list of "subjects" that are assigned the roles.
+//
+// A subject can be a "user" or a "group", or represented with a set of attributes. Specifically,
```
I think we need a really user friendly docs writeup of how this is all supposed to tie together. Thinking about this, and not knowing much about RBAC beyond the textbook parts, I'm guessing this all works because we can identify subjects via client certificates. How would this function without client certs? I think it would be helpful to explain overall context, provide an architecture diagram, explain how it works specifically for the kind of setup we have in places like Istio with mTLS, as well as how it can be adapted.

mattklein123 replied:
+1 (my other comments)

liminw replied:
I added a README.md. Please take a look to see it answers all the questions here.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+//   * User (service account) "cluster.local/ns/default/sa/bookinfo-reviews"
+//   * Service "reviews.default.svc.cluster.local" at version "v1".
+//
+//     kind: ServiceRoleBinding
```
Where does this kind/metadata stuff come from?

liminw replied:
This is the case that ServiceRole and ServiceRoleBinding expressed in K8S CRD format. I explained it in the new README.md.

"The storage of ServiceRole and ServiceRoleBinding objects are platform specific. For example,
 In [Istio](https://istio.io/docs/concepts/security/rbac.html), they are stored as [Kubernetes
 CustomResourceDefinition (CRD)](https://kubernetes.io/docs/concepts/api-extension/custom-resources/)
 objects. In the following document, we will see ServiceRole and ServiceRoleBinding examples
 expressed in Kubernetes CRD format."

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+//       - attributes:
+//           source.service: "reviews.default.svc.cluster.local"
+//           source.labels["version"]: "v1"
+//       roleRef:
```
Please be consistent and use role_ref. We don't use the case switching in Envoy when going from proto name field to JSON/YAML.

liminw replied:
Fixed.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+
+// ServiceRoleBinding assigns a ServiceRole to a list of subjects.
+// This represents the "Spec" part of the ServiceRoleBinding object. The name and namespace
+// of the ServiceRoleBinding is specified in "metadata" section of the ServiceRoleBinding
```
I don't see where metadata is..

liminw replied:
Removed this comment. "metadata" is really platform specific storage of ServiceRole/ServiceRoleBinding objects.

htuch commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+  // "/books/review" (exact match), or "/books/*" (prefix match),
+  // or "*/review" (suffix match).
+  // If not specified, it applies to any path.
+  repeated string paths = 2;
```
Do you want to perhaps use a `RouteMatch` instead? https://github.com/envoyproxy/data-plane-api/blob/f88047c0d470352ad1bf6a97dd64155bc65047a0/envoy/api/v2/route/route.proto#L209

liminw replied:
The match applies to "services", "paths", "constraints" values. I added a StringMatch type, which is similar to RouteMatch.

mattklein123 commented on envoy/config/rbac/v2alpha/local_rbac.proto:
```diff
@@ -0,0 +1,63 @@
+//       }
+//     },
+//  }
+message LocalRBAC {
```
I still honestly don't really understand the concepts here. Can you define before the example:
- Fully qualified service name
- Service policy
- RoleInfo
- ServiceRole
- ServiceRoleBindings

Basically, here is my 99% use case: I want to add some RBAC policies on a route/vhost, or that match various things at the global request level. I don't really care about services, roles, etc. Can you clarify how I accomplish my very basic case and then build from there?

At the documentation level, I still am confused on how "role" is actually going to be used. How will the filter actually operate on this information?

wora replied:
I agree with Matt. We seem to introduce too many concepts for basic things. For most people, they just need let x do y. It can be as simple as:
```
- paths: ["/products/*"]
  users: ["alice@yahoo.com"]
```

mattklein123 commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+
+package envoy.config.rbac.v2alpha;
+
+// RBAC (Role Based Access Control) defines ServiceRole and ServiceRoleBinding
```
Hmm maybe in this section you are answering my questions. Can you merge the docs into one place and refer to the other place if needed?

mattklein123 commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+  // Definition of a custom constraint. The key in the constraint refers to an attribute defined in
+  // AttributeContext (//envoy/service/auth/v2/attribute_context.proto). For example,
+  // destination.labels["app"], request.http.headers["version"].
+  message Constraint {
```
There should be no string based constraints in here at all. IMO everything should be fully specified in protos. Otherwise the proxy will just have to translate strings to real things, which seems kind of silly. Let's just fully define the constraints we want, and then we can always add more later. Optimally you can use AttributeContext directly...

wora replied:
RBAC should be modeled as "pattern matching on AttributeContext". The above comment seems to show the key refers to an attribute.

mattklein123 commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+
+  // Optional. Extra constraints in the ServiceRole specification.
+  // The above ServiceRole examples shows an example of constraint "version".
+  repeated Constraint constraints = 4;
```
I feel pretty strongly that we should allow both whitelist and blacklist modes, as its inevitably going to be asked for. Can we do something similar to the auth proto that we deleted?

wora replied:
We need to decide how to pattern matching the attributes in an efficient way. Otherwise, the performance can be terrible. IMO, Envoy needs to process 10k requests per second, so 100us per request. We must have a performance goal when designing an infrastructure.

liminw commented:
@mattklein123 @htuch I will add a documentation of the end-to-end RBAC flow (from user input -> control plane -> enforcement at data plane). We are missing the architecture view of RBAC. That's where the main confusion is from, I think.

wora commented on envoy/config/rbac/v2alpha/local_rbac.proto:
```diff
@@ -0,0 +1,63 @@
+
+import "envoy/config/rbac/v2alpha/rbac.proto";
+
+// LocalRBAC specifies RBAC policies that apply to the local RBAC engine
```
Do we really need to know whether the RBAC is local or remote? Who should be reader of this info?

liminw replied:
"local" means the RBAC filter spec seen by the proxy. I explained it in README.md.

wora commented on envoy/config/rbac/v2alpha/rbac.proto:
```diff
@@ -0,0 +1,155 @@
+
+// AccessRule defines a permission to access a list of services.
+message AccessRule {
+  // Required. A list of service names.
```
Where is the service name defined? Envoy is not Istio. In most case, Envoy seems an HTTP request, which consists of a URL plus a few headers. How does Envoy map a request to a service?

liminw replied:
service is mapped to "destination.service" attribute.

liminw commented:
@wora Thanks for reviewing the change. Regarding your comments on AWS IAM. AWS IAM, as I understand, is for access control for AWS resources. RBAC we introduced here is to provide RPC (method) level access control for any custom services. They are designed for different goals. I added the "goals" and "backgrounds" part in the change description. And also added a README.md file. I hope it will help clarify a little.

liminw commented:
@mattklein123 @htuch @wora @lizan @rodaine I added a README.md, and made some minor changes to the proto. Could you please take a look to see if your comments are addressed? Thanks!

mattklein123 commented:
@liminw I'm sorry but I think we are going to need to have a meeting. Can we set one up? If anyone out there wants to join can you speak up?

I think the fundamental problem I'm having here is these protos are trying to do two things:
1) Describe control plane storage for  RBAC policies.
2) Describe data plane/runtime implementation of RBAC policies.

I think these two things need to be split, and I don't think (1) needs to live in data-plane-api. 

Essentially, from my perspective https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/auth/auth.proto is all we need (plus some additions). We can use this to describe matching policies that lead to approve/deny decisions. The control plane can generate simple configurations in whatever way it wants.

My vote is essentially to really focus on the split, and look at https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/auth/auth.proto as a starting point for (2) above. How does it fall short for (2)? Do we need (1) in this repo at all?

htuch commented:
Please include me in the meeting invite @liminw.

wora commented:
Please include me as well.

mattklein123 commented:
@liminw given the repo reorg happening this weekend, and the fact that I think we agree this PR is not going to merge in this form, I'm going to go ahead and close. Please open a new Envoy side PR next week and reference this PR for prior discussion. Thank you!
